README


This archive contains the distribution App-Cronjob,
version 1.200004:

  wrap up programs to be run as cron jobs

This software is copyright (c) 2009 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.025.



Changes
Revision history for App-Cronjob

1.200004  2014-11-18 17:58:57-05:00 America/New_York
        - --temp-ignore-lock-errors=3600 now lets you ignore lock errors
          iff the lock is 1h old or less

1.200003  2014-01-06 19:08:15 America/New_York
        - Process::Status is now used to interpret $?, making reports
          more human-oriented and less computer-oriented

1.200002  2013-12-07 08:52:23 America/New_York
        - do not log the pid to syslog; it should already be there
        - if a jobname was given, use it as logging prefix

1.200001  2013-06-19 21:56:28 America/New_York
          update bugtracker

1.200000  2013-04-28 19:43:58 America/New_York
          locking is now done with flock, not O_CREAT|O_EXCL (much of this
          work by Mark Jason Dominus)

          there are now two error types: lockfile (can't create the file) and
          lock (can't get a lock on it)

1.102311  2010-11-02 17:09:23 America/New_York
          lockfile now includes cronjob's pid
          "can't locked, already locked" message now includes lockfile name

1.102310  2010-08-19 12:40:51 America/New_York
          add an Auto-Submitted: auto-generated header (RFC 3834)
          documentation typos fixed (thanks, chromatic)

1.101660  2010-06-15 17:01:46 America/New_York
          added some more documentation to cronjob

1.101230  2010-05-03 17:32:03 America/New_York
          allow A-Z and _ in jobname (requested by AVAR

1.100080  2010-01-08 08:30:52 America/New_York
          syslog when job completes (requested by Bryan Allen)

1.100070  2010-01-07 14:30:09 America/New_York
          switch to Email::Sender::Simple
          improve locking: do not delete lockfile if we didn't lock
          (thanks to sungo and chargrill)

1.093480  2009-12-14 22:15:18 America/New_York
          add prereq for IPC::Run3
          include a pointer, in the App::Cronjob docs, to the bin/cronjob app

1.003     2009-04-21
          fix a moronic bug whereby -> closed a q< string

1.002     2009-04-21
          moved logic to App::Cronjob (not reusable yet; awful)
          fix bug where output caused "FAIL" label instead of just mail

1.000     2009-03-08
          first release with Pobox-specific guts replaced with free guts


lib/App/Cronjob.pm
use strict;
use warnings;
package App::Cronjob;
# ABSTRACT: wrap up programs to be run as cron jobs
$App::Cronjob::VERSION = '1.200004';
use Digest::MD5 qw(md5_hex);
use Errno;
use Fcntl qw( :DEFAULT :flock );
use Getopt::Long::Descriptive;
use IPC::Run3 qw(run3);
use Log::Dispatchouli;
use Process::Status 0.002;
use String::Flogger;
use Sys::Hostname::Long;
use Text::Template;
use Time::HiRes ();

#pod =head1 SEE INSTEAD
#pod
#pod This library, App::Cronjob, is not well documented.  Its internals may change
#pod substantially until such point as it is documented.
#pod
#pod Instead of using the library, you should run the program F<cronjob> that is
#pod installed along with the library.
#pod
#pod For a full description of the program's behavior, consult L<cronjob>.
#pod
#pod =cut

my $TEMPLATE;

my (
  $opt,
  $usage,
  $subject,
  $rcpts,
  $host,
  $sender,
);

sub run {
  ($opt, $usage) = describe_options(
    '%c %o',
     [ 'command|c=s',   'command to run (passed to ``)', { required => 1 }   ],
     [ 'subject|s=s',   'subject of mail to send (defaults to command)'      ],
     [ 'rcpt|r=s@',     'recipient of mail; may be given many times',        ],
     [ 'errors-only|E', 'do not send mail if exit code 0, even with output', ],
     [ 'sender|f=s',    'sender for message',                                ],
     [ 'jobname|j=s',   'job name; used for locking, if given'               ],
     [ 'ignore-errors=s@', 'error types to ignore (like: lock)'              ],
     [ 'temp-ignore-lock-errors=i',
                     'failure to lock only signals an error after this long' ],
     [ 'lock!',         'lock this job (defaults to true; --no-lock for off)',
                        { default => 1 }                                     ],
  );

  $subject = $opt->{subject} || $opt->{command};
  $subject =~ s{\A/\S+/([^/]+)(\s|$)}{$1$2} if $subject eq $opt->{command};

  if (defined $opt->{temp_ignore_lock_errors}) {
    if (grep {; $_ eq "lock" } @{$opt->{ignore_errors}}) {
      die "--temp-ignore-lock-errors and --ignore-errors=lock are incompatible\n";
    }
  }

  $rcpts   = $opt->{rcpt}
          || [ split /\s*,\s*/, ($ENV{MAILTO} ? $ENV{MAILTO} : 'root') ];

  $host    = hostname_long;
  $sender  = $opt->{sender} || sprintf '%s@%s', ($ENV{USER}||'cron'), $host;

  my $lockfile = sprintf '/tmp/cronjob.%s',
                 $opt->{jobname} || md5_hex($subject);

  my $got_lock;

  my $okay = eval {
    die "illegal job name: $opt->{jobname}\n"
      if $opt->{jobname} and $opt->{jobname} !~ m{\A[-_A-Za-z0-9]+\z};

    my $logger  = Log::Dispatchouli->new({
      ident    => 'cronjob',
      facility => 'cron',
      log_pid  => 0,
      (defined $opt->{jobname} ? (prefix => "$opt->{jobname}: ") : ()),
    });

    my $lock_fh;
    if ($opt->lock) {
      sysopen $lock_fh, $lockfile, O_CREAT|O_WRONLY
        or die App::Cronjob::Exception->new(
          lockfile => "couldn't open lockfile $lockfile: $!"
        );

      my $lock_flags = LOCK_EX | LOCK_NB;

      unless (flock $lock_fh, $lock_flags) {
        my $mtime = (stat $lock_fh)[9];
        my $stamp = scalar localtime $mtime;
        die App::Cronjob::Exception->new(
          lock => "can't lock; locked since $stamp",
          { locked_since => $mtime },
        );
      }

      printf $lock_fh "running %s\nstarted at %s\n",
        $opt->{command}, scalar localtime $^T;

      $got_lock = 1;
    }

    $logger->log([ 'trying to run %s', $opt->{command} ]);

    my $start = Time::HiRes::time;
    my $output;

    # XXX: does not throw proper exception
    $logger->log_fatal([ 'run3 failed to run command: %s', $@ ])
      unless eval { run3($opt->{command}, \undef, \$output, \$output); 1; };

    my $status = Process::Status->new;

    my $end = Time::HiRes::time;

    my $send_mail = ($status->exitstatus != 0)
                 || (length $output && ! $opt->{errors_only});

    my $time_taken = sprintf '%0.4f', $end - $start;

    $logger->log([
      'job completed with status %s after %ss',
      $status->as_struct,
      $time_taken,
    ]);

    if ($send_mail) {
      send_cronjob_report({
        is_fail => (!! $status->exitstatus),
        status  => $status,
        time    => \$time_taken,
        output  => \$output,
      });
    }

    1;
  };

  exit 0 if $okay;
  my $err = $@;

  if (eval { $err->isa('App::Cronjob::Exception'); }) {
    unless (
      grep { $err->{type} and $_ eq $err->{type} } @{$opt->{ignore_errors}}
    ) {
      if ($err->{type} eq "lock" && $opt->{temp_ignore_lock_errors}) {
        my $age = time() - $err->{extra}{locked_since};
        exit 0 if $age <= $opt->{temp_ignore_lock_errors};
      }
      send_cronjob_report({
        is_fail => 1,
        output  => \$err->{text},
      });
    }

    exit 0;
  } else {
    $subject = "ERROR: $subject";
    send_cronjob_report({
      is_fail => 1,
      output  => \$err
    });
    exit 0;
  }
}

# read INI from /etc/cronjob
#sub __config {
#}

sub send_cronjob_report {
  my ($arg) = @_;

  require Email::Simple;
  require Email::Simple::Creator;
  require Email::Sender::Simple;
  require Text::Template;

  my $body     = Text::Template->fill_this_in(
    $TEMPLATE,
    HASH => {
      command => \$opt->{command},
      output  => $arg->{output},
      time    => $arg->{time} || \'(n/a)',
      status  => \($arg->{status} ? $arg->{status}->as_string : 'never ran'),
    },
  );

  my $subject = sprintf '%s%s', ($arg->{is_fail} ? 'FAIL: ' : ''), $subject;

  my $irt = sprintf '<%s@%s>', md5_hex($subject), $host;

  my $email = Email::Simple->create(
    body   => $body,
    header => [
      To      => join(', ', @$rcpts),
      From    => qq{"cron/$host" <$sender>},
      Subject => $subject,
      'In-Reply-To' => $irt,
      'Auto-Submitted' => 'auto-generated',
    ],
  );

  Email::Sender::Simple->send(
    $email,
    {
      to      => $rcpts,
      from    => $sender,
    }
  );
}

BEGIN {
$TEMPLATE = <<'END_TEMPLATE'
Command: { $command }
Time   : { $time }s
Status : { $status }

Output :

{ $output || '(no output)' }
END_TEMPLATE
}

{
  package App::Cronjob::Exception;
$App::Cronjob::Exception::VERSION = '1.200004';
sub new {
    my ($class, $type, $text, $extra) = @_;
    bless { type => $type, text => $text, extra => $extra } => $class;
  }
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::Cronjob - wrap up programs to be run as cron jobs

=head1 VERSION

version 1.200004

=head1 SEE INSTEAD

This library, App::Cronjob, is not well documented.  Its internals may change
substantially until such point as it is documented.

Instead of using the library, you should run the program F<cronjob> that is
installed along with the library.

For a full description of the program's behavior, consult L<cronjob>.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.025.
use strict;
use warnings;



use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "wrap up programs to be run as cron jobs",
  "AUTHOR" => "Ricardo Signes <rjbs\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-Cronjob",
  "EXE_FILES" => [
    "bin/cronjob"
  ],
  "LICENSE" => "perl",
  "NAME" => "App::Cronjob",
  "PREREQ_PM" => {
    "Digest::MD5" => 0,
    "Email::Sender::Simple" => "0.091661",
    "Email::Simple" => 0,
    "Email::Simple::Creator" => 0,
    "Fcntl" => 0,
    "Getopt::Long::Descriptive" => 0,
    "IPC::Run3" => 0,
    "Log::Dispatchouli" => 0,
    "Process::Status" => "0.002",
    "String::Flogger" => 0,
    "Sys::Hostname::Long" => 0,
    "Text::Template" => 0,
    "Time::HiRes" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "1.200004",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Digest::MD5" => 0,
  "Email::Sender::Simple" => "0.091661",
  "Email::Simple" => 0,
  "Email::Simple::Creator" => 0,
  "ExtUtils::MakeMaker" => 0,
  "Fcntl" => 0,
  "File::Spec" => 0,
  "Getopt::Long::Descriptive" => 0,
  "IPC::Run3" => 0,
  "Log::Dispatchouli" => 0,
  "Process::Status" => "0.002",
  "String::Flogger" => 0,
  "Sys::Hostname::Long" => 0,
  "Test::More" => "0.96",
  "Text::Template" => 0,
  "Time::HiRes" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



